### PR TITLE
[SERVICE-282] Add additional test coverage to Save and Restore

### PR DESCRIPTION
**Pull request by MichaelMCoates**
- Adds S&R testing for applications created from a manifest
- Adds S&R testing for deregistered applications
- Removes tests with 3 children to save time and limit likelihood of tests failing for windows off-screen

### CODE REVIEW

File: test/demo/utils/serviceUtils.ts [hunk 1]
```diff
@@ -46,6 +44,12 @@ export async function sendServiceMessage<T, R>(message: string, payload: T): Pro
     return client.dispatch(message, payload);
 }
 
+
+export async function sendServiceMessageAwait<T, R>(message: string, payload: T): Promise<R> {
```
Comment (pjbroadbent) [inline test/demo/utils/serviceUtils.ts]: This is functionally identical to `sendServiceMessage`, can replace any uses of this method with the existing one.

File: test/demo/utils/workspacesUtils.ts [hunk 2]
```diff
@@ -53,27 +52,41 @@ export async function createCloseAndRestoreLayout(t: SaveRestoreTestContext) {
...
     await Promise.all(t.context.testAppData.map(async (appData: TestAppData) => await appData.app.close(true)));
     await assertAllAppsClosed(t);
-    await sendServiceMessage('restoreLayout', generatedLayout);
+    await sendServiceMessageAwait('restoreLayout', generatedLayout);
+    // To give placeholder windows time to disappear.
+    // The tests close out all testing windows upon completion.
+    // The placeholders listens to the show-requested event of its testing window, moves the window, shows it, and closes itself.
+    // If the tests run too quickly after restore, the placeholder windows may receive the child window show event too late,
+    // and the window may not exist, which makes the placeholder windows stay open.
+    await delay(500);
```
Comment (pjbroadbent) [inline test/demo/utils/workspacesUtils.ts]: We'll ultimately need to handle this within the provider (`restoreLayout` shouldn't resolve until all placeholders have been cleared up), but can treat that as tech debt.

File: test/demo/utils/AppInitializer.ts [hunk 3]
```diff
@@ -54,21 +54,27 @@ const OPTIONS_BASE = {
...
-const APP_INITIALIZER_BASE = {
+const APP_INITIALIZER_BASE_PROGRAMMATIC = {
     appOptions: OPTIONS_BASE,
-    createType: 'programmatic',
+    createType: 'programmatic' as 'programmatic',
     childWindows: []
 };
 
+const APP_INITIALIZER_BASE_MANIFEST = {
+    createType: 'manifest' as 'manifest',
```
Comment (pjbroadbent) [inline test/demo/utils/AppInitializer.ts]: If you add type definitions to these constants, the casts aren't necessary. Also (IMO) makes it clearer as to what these do.

e.g:
```ts
const APP_INITIALIZER_BASE_MANIFEST: ManifestTestParams = {
    createType: 'manifest',
    // ...
```

File: test/runner.js [hunk 4]
```diff
@@ -134,6 +134,31 @@ async function serve() {
 
         app.use(express.static('dist'));
         app.use(express.static('res'));
+
+        // Add route to dynamically generate app manifests
+        app.use('/create-manifest', (() => (req, res) => {
```
Comment (pjbroadbent) [inline test/runner.js]: Does this need to be an IIFE? Could be `app.use('/create-manifest', (req, res) => {`, like the handler above?

File: test/demo/utils/serviceUtils.ts (hunk 1 above)
Comment (MichaelMCoates) [reply test/demo/utils/serviceUtils.ts]: Wow, yeah, not sure what I was doing with this.

File: test/demo/utils/AppInitializer.ts (hunk 3 above)
Comment (MichaelMCoates) [reply test/demo/utils/AppInitializer.ts]: Ah right, thanks!

File: test/demo/utils/workspacesUtils.ts [hunk 5]
```diff
@@ -53,27 +52,41 @@ export async function createCloseAndRestoreLayout(t: SaveRestoreTestContext) {
...
     return windowGroupings.map(windowGrouping => {
         return {apps: appsArray as AppInitializerParams[], snapWindowGrouping: windowGrouping};
     });
 }
 
-export function createTabTests(numApps: number, children: number): {apps: AppInitializerParams[], tabWindowGrouping: WindowGrouping}[] {
-    const windowGroupings = createWindowGroupings(numApps, children);
-    const appsArray = createAppsArray(numApps, children);
+export function createTabTests(
+    numAppsToCreate: number, numberOfChildren: number, testOptions?: TestCreationOptions): {apps: AppInitializerParams[], tabWindowGrouping: WindowGrouping}[] {
```
[the remark this answers is not in the document]
Comment (MichaelMCoates) [reply test/demo/utils/workspacesUtils.ts]: Gotcha. Also lets me get rid of the casts.

File: test/runner.js (hunk 4 above)
Comment (MichaelMCoates) [reply test/runner.js]: Totally right, fixed.

File: test/demo/utils/workspacesUtils.ts (hunk 2 above)
Comment (MichaelMCoates) [reply test/demo/utils/workspacesUtils.ts]: Sounds good.

**Comment by tomrobertsOF** [comment]
Can you re-lint this. It's failing in Jenkins.

**Comment by MichaelMCoates** [comment]
Linted!

**Comment by tomrobertsOF** [comment]
Cool! Since it's adding tests I just want to make sure it runs properly in CI before we merge.

**Comment by MichaelMCoates** [comment]
^Rebase

**Comment by MichaelMCoates** [comment]
Rebase ^

**Pull request merged**